spec/system/admin/order_spec.rb

```diff
@@ -127,6 +127,34 @@ def new_order_with_distribution(distributor, order_cycle)
     expect(page).not_to have_select2 "add_variant_id", with_options: [product.name]
   end
 
+  context "cancelling an order" do
+    let(:line_item) { create(:line_item) }
+
+    before do
+      order.line_items << line_item
+      login_as_admin_and_visit spree.edit_admin_order_path(order)
+    end
+
+    context "when using the cancel button" do
+      it "shows a modal about order cancellation" do
+        find("#cancel_order_form").click
+        expect(page).to have_content "This will cancel the current order."
+        expect(page).to have_checked_field "Send a cancellation email to the customer"
+        expect(page).to have_checked_field "Restock Items: return all items to stock"
+      end
+    end
+
+    context "when using the cancel option in the dropdown" do  
+      it "shows a modal about order cancellation" do
+        find("#links-dropdown .ofn-drop-down").click
+        find('a[href$="cancel"]').click
+        expect(page).to have_content "This will cancel the current order."
+        expect(page).to have_checked_field "Send a cancellation email to the customer"
+        expect(page).to have_checked_field "Restock Items: return all items to stock"
+      end
+    end
```

Comment:
To factorize spec, you could use this kind of syntax:

```ruby
    shared_examples_for "clicking the cancel link" do
      it "shows a modal about order cancellation" do
        expect(page).to have_content "This will cancel the current order."
        expect(page).to have_checked_field "Send a cancellation email to the customer"
        expect(page).to have_checked_field "Restock Items: return all items to stock"
      end
    end

    it_should_behave_like "clicking the cancel link" do
      before { find("#cancel_order_form").click }
    end

    it_should_behave_like "clicking the cancel link" do
      before do
        find("#links-dropdown .ofn-drop-down").click
        find('a[href$="cancel"]').click
      end
    end
```

Also I would add some spec that test the `Cancel` and `Ok` button on modal:
```ruby
      it "cancels the order" do
        within ".modal" do
          expect do
            click_button "Cancel Order"
            expect(page).to have_content "Order cancelled"
          end.to change { order.reload.state }.from("cart").to("canceled")
        end
        expect(page).to have_content "Order cancelled"
      end
      it "close the modal when clicking on the close button" do
        within ".modal" do
          expect do
            click_button "Close"
          end.not_to change { order.reload.state }
        end
      end
```

(this is pseudo code, not tested)

Reply:
Thanks, I have a question about the last part, in the same file there is a spec that tests the modal because the modal also appears when deleting the last item, and the spec tests both the cancel and ok, wouldn't I just be testing the same functionality again?

https://github.com/openfoodfoundation/openfoodnetwork/blob/7b14afbabedbe958232c44013b636996b7b3f0f6/spec/system/admin/order_spec.rb#L169

Reply:
Good question. It's actually the "same functionality", but not triggered by the same context (deleting the last item vs. clicking on cancel). As we are testing in _Integration_ level, I would say that we should test same functionality in different contexts, to be as close as the user. For sure, as this is the same modal we could factorize the spec as well.

Maybe @filipefurtad0 has a point of view about it.

Reply:
Hey @basilawwad ,
Thanks for your work on this so far :pray: 

> It's actually the "same functionality", but not triggered by the same context (deleting the last item vs. clicking on cancel)

:100: agree. That context block should be shared throughout all the order cancelling scenarios:
- editing the order and i) clicking `Cancel` or ii) selecting `Cancel` from the dropdown in the orders page (the contexts on lines :138 and :147, respectively)
- deleting item of an order (context in line :158) -> done

So, and has @jibees suggests, a way to go around this would be to implement the shared example throughout these three test cases :+1: 

PS: and note to myself - we should also have this shared example on `./spec/system/admin/orders_spec.rb`:
- (selecting `Cancel` from the dropdown in the `/orders` page (after ticking that order) -> which actually relates to the recent PR https://github.com/openfoodfoundation/openfoodnetwork/pull/10275/files)